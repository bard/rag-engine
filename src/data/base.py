import hashlib
from typing import Self
from langchain_core.language_models import BaseChatModel
from pydantic import BaseModel


class IndexableData(BaseModel):
    title: str
    source_url: str

    def id(self) -> str:
        content_hash = hashlib.sha256(self.to_text().encode("utf-8")).hexdigest()
        # TODO consider using a url-based id (allows replacing/updating as long as urls are stable)
        return f"{self.__class__.__name__}-{content_hash[:8]}"

    def to_text(self) -> str:
        raise Exception("to_text() not defined")

    @classmethod
    def from_html(
        cls, html: str, source_url: str, llm: BaseChatModel | None
    ) -> Self | None:
        raise Exception("from_html() not defined")
