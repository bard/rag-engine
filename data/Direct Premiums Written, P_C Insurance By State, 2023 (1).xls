
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  </head>
  <body>
    <table>
    <thead><tr><th></th></tr></thead>    <tbody> <tr class="odd"><td><table>
<tr><td>Direct Premiums Written, P/C Insurance By State, 2023 (1)</td></tr>
<tr><td><p>($000)</p>
</td></tr>
<tr><td><table>
<thead>
<tr>
<th>State</th>
<th>Total, all lines</th>
<th>State</th>
<th>Total, all lines</th>
</tr>
</thead>
<tbody>
<tr>
<td>Alabama</td>
<td>$13,146,853</td>
<td>Montana</td>
<td>$3,807,820</td>
</tr>
<tr>
<td>Alaska</td>
<td>1,994,750</td>
<td>Nebraska</td>
<td>7,573,101</td>
</tr>
<tr>
<td>Arizona</td>
<td>17,959,495</td>
<td>Nevada</td>
<td>8,561,822</td>
</tr>
<tr>
<td>Arkansas</td>
<td>7,931,849</td>
<td>New Hampshire</td>
<td>3,342,091</td>
</tr>
<tr>
<td>California</td>
<td>111,243,502</td>
<td>New Jersey</td>
<td>29,106,640</td>
</tr>
<tr>
<td>Colorado</td>
<td>20,582,327</td>
<td>New Mexico</td>
<td>5,038,116</td>
</tr>
<tr>
<td>Connecticut</td>
<td>11,508,412</td>
<td>New York</td>
<td>61,861,773</td>
</tr>
<tr>
<td>D.C.</td>
<td>2,699,081</td>
<td>North Carolina</td>
<td>24,036,512</td>
</tr>
<tr>
<td>Delaware</td>
<td>3,842,583</td>
<td>North Dakota</td>
<td>3,826,625</td>
</tr>
<tr>
<td>Florida</td>
<td>88,067,510</td>
<td>Ohio</td>
<td>23,397,453</td>
</tr>
<tr>
<td>Georgia</td>
<td>32,039,933</td>
<td>Oklahoma</td>
<td>11,325,023</td>
</tr>
<tr>
<td>Hawaii</td>
<td>3,490,752</td>
<td>Oregon</td>
<td>10,510,197</td>
</tr>
<tr>
<td>Idaho</td>
<td>4,757,044</td>
<td>Pennsylvania</td>
<td>32,532,786</td>
</tr>
<tr>
<td>Illinois</td>
<td>35,938,656</td>
<td>Rhode Island</td>
<td>3,361,749</td>
</tr>
<tr>
<td>Indiana</td>
<td>16,321,562</td>
<td>South Carolina</td>
<td>15,078,257</td>
</tr>
<tr>
<td>Iowa</td>
<td>9,843,752</td>
<td>South Dakota</td>
<td>3,970,217</td>
</tr>
<tr>
<td>Kansas</td>
<td>9,565,831</td>
<td>Tennessee</td>
<td>17,442,243</td>
</tr>
<tr>
<td>Kentucky</td>
<td>10,147,626</td>
<td>Texas</td>
<td>92,331,270</td>
</tr>
<tr>
<td>Louisiana</td>
<td>16,226,096</td>
<td>Utah</td>
<td>8,100,467</td>
</tr>
<tr>
<td>Maine</td>
<td>3,316,037</td>
<td>Vermont</td>
<td>1,640,326</td>
</tr>
<tr>
<td>Maryland</td>
<td>16,186,978</td>
<td>Virginia</td>
<td>20,316,519</td>
</tr>
<tr>
<td>Massachusetts</td>
<td>21,453,245</td>
<td>Washington</td>
<td>18,109,893</td>
</tr>
<tr>
<td>Michigan</td>
<td>24,427,664</td>
<td>West Virginia</td>
<td>3,643,445</td>
</tr>
<tr>
<td>Minnesota</td>
<td>17,081,121</td>
<td>Wisconsin</td>
<td>14,623,479</td>
</tr>
<tr>
<td>Mississippi</td>
<td>7,505,441</td>
<td>Wyoming</td>
<td>1,820,520</td>
</tr>
<tr>
<td>Missouri</td>
<td>17,261,874</td>
<td><strong>United States</strong></td>
<td><strong>$949,898,320</strong></td>
</tr>
</tbody></table>

(1) Before reinsurance transactions, includes state funds, excludes territories.
(2) Data for the total United States may differ from similar data shown elsewhere due to the use of different exhibits from S&amp;P Global Market Intelligence.

Source: NAIC data, sourced from S&amp;P Global Market Intelligence, Insurance Information Institute.</td></tr>
</table></td> </tr>
      </tbody>
    </table>
  </body>
</html>