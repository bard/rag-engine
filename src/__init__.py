from . import ingest
from . import query
