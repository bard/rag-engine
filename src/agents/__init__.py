from .document_processor import DocumentProcessor
from .vector_store import VectorStore

__all__ = ["DocumentProcessor", "VectorStore"]
