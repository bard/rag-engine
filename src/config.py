from langchain_core.runnables.config import RunnableConfig
from pydantic import BaseModel, Field
from typing import Self, Union, Literal


class PineconeVectorStoreBackend(BaseModel):
    type: Literal["pinecone"]
    index_name: str
    api_key: str
    embeddings_type: Literal["openai"]
    score_threshold: float = 0.1


class MockVectorStoreBackend(BaseModel):
    type: Literal["mock"]
    score_threshold: float = 0.1


class ChromaVectorStoreBackend(BaseModel):
    type: Literal["chroma"]
    path: str
    collection_name: str
    embeddings_type: Literal["local-minilm", "openai"]
    score_threshold: float = 0.1


type VectorStoreBackend = Union[
    PineconeVectorStoreBackend, MockVectorStoreBackend, ChromaVectorStoreBackend
]


class OpenaiLlmBackend(BaseModel):
    type: Literal["openai"]
    model: str
    api_key: str


class AnthropicLlmBackend(BaseModel):
    type: Literal["anthropic"]
    model: str
    api_key: str


class DbBackend(BaseModel):
    url: str


type LlmBackend = Union[OpenaiLlmBackend, AnthropicLlmBackend]


class AgentConfig(BaseModel):
    db: DbBackend
    llm: LlmBackend = Field(discriminator="type")
    vector_store: VectorStoreBackend = Field(discriminator="type")

    @classmethod
    def from_runnable_config(cls, config: RunnableConfig) -> Self:
        """Validate runnable config"""
        configurable = config.get("configurable")
        assert configurable is not None
        return cls(**configurable)
