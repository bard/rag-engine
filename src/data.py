from langchain.schema import Document
from typing import Dict, Any, Self
from pydantic import BaseModel
from bs4 import BeautifulSoup
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase


class InsuranceRecord(BaseModel):
    """Pydantic model for insurance record data"""

    year: int
    average_expenditure: float
    percent_change: float
    source_content: str

    def to_langchain_document(self, source_url: str) -> Document:
        """
        Convert this InsuranceRecord to a LangChain Document.

        Args:
            source_url: The URL where the record was sourced from

        Returns:
            Document object containing the record data and metadata
        """
        content = "\n".join(
            [
                f"Year: {self.year}, "
                f"Expenditure: ${self.average_expenditure:.2f}, "
                f"Change: {self.percent_change}%"
            ]
        )

        return Document(
            page_content=content,
            id=f"insurance-record-{self.year}",
            metadata={
                "source_url": source_url,
                "source_content": self.source_content,
                "year": self.year,
                "average_expenditure": self.average_expenditure,
                "percent_change": self.percent_change,
            },
        )

    @classmethod
    def from_html_content(cls, html_content: str) -> list[Self]:
        """
        Parse the auto insurance expenditure table and return a list of dictionaries.
        Each dictionary represents a row with year, expenditure, and percent change.
        """
        # Create BeautifulSoup object
        soup = BeautifulSoup(html_content, "html.parser")

        # Find the innermost table with the actual data
        tables = soup.find_all("table")
        data_table = tables[2]  # The third table contains our data

        # Initialize list to store results
        results = []

        # Process each row in tbody
        for row in data_table.tbody.find_all("tr"):
            cols = row.find_all("td")

            # Extract values
            year = int(cols[0].text.strip())

            # Clean and convert expenditure to float
            expenditure = cols[1].text.strip()
            expenditure = expenditure.replace("$", "").replace(",", "")
            expenditure = float(expenditure)

            # Clean and convert percent change to float
            percent_change = cols[2].text.strip()
            percent_change = percent_change.replace("%", "")
            percent_change = float(percent_change)

            # Create dictionary for current row
            record = cls(
                year=year,
                average_expenditure=expenditure,
                percent_change=percent_change,
                source_content=str(row),
            )
            results.append(record)

        return results


class SqlAlchemyBase(DeclarativeBase):
    pass


class SQLInsuranceRecord(SqlAlchemyBase):
    """SQLAlchemy model for insurance record data"""

    __tablename__ = "insurance_records"

    id: Mapped[int] = mapped_column(primary_key=True)
    year: Mapped[int] = mapped_column(nullable=False)
    average_expenditure: Mapped[float] = mapped_column(nullable=False)
    percent_change: Mapped[float] = mapped_column(nullable=False)
    source_content: Mapped[str] = mapped_column(nullable=False)

    def __repr__(self) -> str:
        return (
            f"SQLInsuranceRecord(id={self.id}, "
            f"year={self.year}, "
            f"average_expenditure={self.average_expenditure:.2f}, "
            f"percent_change={self.percent_change})"
        )


class RawGenericTabularData(BaseModel):
    title: str
    data: list[Dict[str, Any]]


class GenericTabularData(RawGenericTabularData):
    content_hash: str

    def to_langchain_document(self) -> Document:
        """
        Convert this GenericTabularData to a LangChain Document.

        Returns:
            Document object containing the data and metadata
        """
        content = f"Title: {self.title}\n"
        for item in self.data:
            content += "\n".join(f"{k}: {v}" for k, v in item.items())
            content += "\n\n"

        return Document(
            page_content=content,
            id=f"generic-data-{self.content_hash[:8]}",
            metadata={
                "title": self.title,
                "content_hash": self.content_hash,
                "raw_data": self.data,
            },
        )
