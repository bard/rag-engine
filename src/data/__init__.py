from .base import IndexableData
from .generic_tabular import GenericTabularData
from .insurance_average_expenditure import InsuranceAverageExpenditureData
from .textual import TextualData

__all__ = [
    "IndexableData",
    "GenericTabularData",
    "InsuranceAverageExpenditureData",
    "TextualData",
]
