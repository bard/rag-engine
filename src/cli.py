import click
import logging
from os import path
from langchain_core.messages import HumanMessage, AIMessage
from sqlalchemy.orm import Session

from src import services, workflow_query, workflow_ingest, db
from src.config import Config
from src.db import SqlTopic


CONFIG = Config.from_env()


@click.group()
@click.option(
    "--log",
    help="Set logging level",
    default="INFO",
    type=click.Choice(
        ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"], case_sensitive=False
    ),
)
def cli(log):
    logging.basicConfig(level=getattr(logging, log.upper()))


@click.command(name="create_topic")
@click.option("--id", required=False, help="Optional topic ID")
@click.option("--name", required=True, help="Topic name")
def cmd_create_topic(id: str | None, name: str):
    db = services.get_db(CONFIG)
    with Session(db) as session:
        topic = SqlTopic(id=id, name=name)
        session.add(topic)
        session.commit()
        session.refresh(topic)

    click.echo()
    click.echo(f"Created topic '{name}' with ID: {topic.id}")
    click.echo()


@click.command(name="ingest")
@click.argument("path_or_url")
@click.option("--topic_id", required=False, help="Optional topic ID")
def cmd_ingest(path_or_url: str, topic_id: str | None):
    if path_or_url.startswith(("http://", "https://", "file://", "data://")):
        input_url = path_or_url
    else:
        abs_path = path.abspath(path_or_url)
        input_url = f"file://{abs_path}"
    initial_agent_state = workflow_ingest.AgentState(
        url=input_url, source_content=None, extracted_data=[], topic_id=topic_id
    )

    workflow_ingest.get_graph().invoke(initial_agent_state, CONFIG.to_runnable_config())

    click.echo()
    click.echo("Data ingested successfully")
    click.echo()


@click.command(name="list_topics")
def cmd_list_topics():
    """List all available topics."""
    db_engine = services.get_db(CONFIG)
    with Session(db_engine) as session:
        topics = session.query(SqlTopic).order_by(SqlTopic.name).all()

    click.echo()
    if topics:
        click.echo("Available topics:")
        for topic in topics:
            click.echo(f"  {topic.id}: {topic.name}")
    else:
        click.echo("No topics found")
    click.echo()


@click.command(name="info")
def cmd_info():
    """Display information about the database and vector store."""
    # Get database stats
    db_engine = services.get_db(CONFIG)
    with Session(db_engine) as session:
        topic_count = session.query(SqlTopic).count()
        doc_count = session.query(db.SqlKnowledgeBaseDocument).count()

    # Get sanitized DB URL (remove credentials)
    db_url = str(CONFIG.db.url)
    if "@" in db_url:
        db_url = db_url.split("@")[1]
        db_url = f"***:***@{db_url}"

    click.echo()
    click.echo("Database Information:")
    click.echo(f"  URL: {db_url}")
    click.echo(f"  Topics: {topic_count}")
    click.echo(f"  Documents: {doc_count}")
    click.echo()


@click.command(name="initdb")
def cmd_initdb():
    engine = services.get_db(CONFIG)

    db.SqlAlchemyBase.metadata.create_all(engine)

    click.echo()
    click.echo("Database initialized successfully")
    click.echo()


@click.command(name="query")
@click.argument("user_query")
@click.option("--topic_id", required=False, help="Optional topic ID")
def cmd_query(user_query: str, topic_id: str | None):
    initial_agent_state = workflow_query.AgentState(
        messages=[HumanMessage(content=user_query)],
        retrieved_knowledge=[],
        query=None,
        topic_id=topic_id,
        external_knowledge_sources=[],
    )

    response = workflow_query.get_graph().invoke(
        initial_agent_state, CONFIG.to_runnable_config()
    )

    ai_messages = [msg for msg in response["messages"] if isinstance(msg, AIMessage)]
    click.echo()
    if ai_messages:
        click.echo(ai_messages[-1].content)
    else:
        click.echo("No AI response found")
    click.echo()


if __name__ == "__main__":
    cli.add_command(cmd_create_topic)
    cli.add_command(cmd_ingest)
    cli.add_command(cmd_info)
    cli.add_command(cmd_initdb)
    cli.add_command(cmd_list_topics)
    cli.add_command(cmd_query)
    cli()
