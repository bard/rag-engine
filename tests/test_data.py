import pytest
import pprint
from src.data import GenericTabularData, ExpenditureReport


def test_parse_average_expenditures_document(
    snapshot, average_insurance_expenditures_html
):
    expenditure_record = ExpenditureReport.from_html_content(
        average_insurance_expenditures_html, source_url="about:blank"
    )

    assert expenditure_record == snapshot


def test_get_langchain_document_from_expenditure_report(
    snapshot, average_insurance_expenditures_html
):
    expenditure_report = ExpenditureReport.from_html_content(
        average_insurance_expenditures_html, source_url="about:blank"
    )

    document = expenditure_report.to_langchain_document()
    assert document == snapshot


def test_get_langchain_document_from_generic_data(
    snapshot,
):
    generic_data = GenericTabularData(
        title="Some Data",
        data=[{"foo": 1, "bar": 2}, {"foo": 4, "bar": 5}],
        source_url="about:blank",
    )

    document = generic_data.to_langchain_document()

    assert document == snapshot
