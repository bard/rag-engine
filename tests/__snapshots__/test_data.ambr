# serializer version: 1
# name: test_get_langchain_document_from_expenditure_report
  Document(id='expenditure-report-710116cb', metadata={'source_url': 'about:blank', 'title': 'Average Expenditures for Auto Insurance, 2012-2021'}, page_content='# Average Expenditures for Auto Insurance, 2012-2021\n\n')
# ---
# name: test_get_langchain_document_from_generic_data
  Document(id='expenditure-report-02794835', metadata={'title': 'Some Data', 'source_url': 'about:blank'}, page_content='# Some Data\n\n')
# ---
# name: test_parse_average_expenditures_document
  ExpenditureReport(title='Average Expenditures for Auto Insurance, 2012-2021', data=[InsuranceRecord(year=2012, average_expenditure=812.4, percent_change=2.2, source_content='<tr>\n<td>2012</td>\n<td>$812.40</td>\n<td>2.2%</td>\n</tr>'), InsuranceRecord(year=2013, average_expenditure=841.06, percent_change=3.5, source_content='<tr>\n<td>2013</td>\n<td>841.06</td>\n<td>3.5</td>\n</tr>'), InsuranceRecord(year=2014, average_expenditure=869.47, percent_change=3.4, source_content='<tr>\n<td>2014</td>\n<td>869.47</td>\n<td>3.4</td>\n</tr>'), InsuranceRecord(year=2015, average_expenditure=896.66, percent_change=3.1, source_content='<tr>\n<td>2015</td>\n<td>896.66</td>\n<td>3.1</td>\n</tr>'), InsuranceRecord(year=2016, average_expenditure=945.22, percent_change=5.4, source_content='<tr>\n<td>2016</td>\n<td>945.22</td>\n<td>5.4</td>\n</tr>'), InsuranceRecord(year=2017, average_expenditure=1008.35, percent_change=6.7, source_content='<tr>\n<td>2017</td>\n<td>1,008.35</td>\n<td>6.7</td>\n</tr>'), InsuranceRecord(year=2018, average_expenditure=1058.1, percent_change=4.9, source_content='<tr>\n<td>2018</td>\n<td>1,058.10</td>\n<td>4.9</td>\n</tr>'), InsuranceRecord(year=2019, average_expenditure=1071.74, percent_change=1.3, source_content='<tr>\n<td>2019</td>\n<td>1,071.74</td>\n<td>1.3</td>\n</tr>'), InsuranceRecord(year=2020, average_expenditure=1046.37, percent_change=-2.4, source_content='<tr>\n<td>2020</td>\n<td>1,046.37</td>\n<td>-2.4</td>\n</tr>'), InsuranceRecord(year=2021, average_expenditure=1061.54, percent_change=1.4, source_content='<tr>\n<td>2021</td>\n<td>1,061.54</td>\n<td>1.4</td>\n</tr>')], source_url='about:blank')
# ---
